[C++23] [CLANG] Adding C++23 constexpr math functions: fmin, fmax and frexp.

zahiraam opened the pull request:
Add support for `fmin`, `fmax`, and (`__builtin_`)`frexp`.

zahiraam committed:
Adding C23 constexpr math functions fmin and frexp.

AaronBallman commented on clang/test/CodeGen/constexpr-math.cpp:
```diff
@@ -0,0 +1,51 @@
+// RUN %clang_cc1 -x c++ -triple x86_64-linux-gnu -emit-llvm -o - %s \
+// RUN -std=c++23
+
+#define INFINITY ((float)(1e+300 * 1e+300))
```
You should also add tests in SemaCXX that the calculations work at compile time. e.g.,
```
static_assert(__builtin_fmin(15.24, 1.3) == 1.3, "");
```

AaronBallman commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14547,6 +14547,18 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
   default:
     return false;
 
+  case Builtin::BI__builtin_frexpf:
```
I think it would be helpful to also update `clang\lib\AST\Interp\InterpBuiltin.cpp` to support the builtins for C++23 constexpr math functions. CC @tbaederr for his opinion.

tbaederr replied:
Not doing that is fine for now.

AaronBallman reviewed (commented):
C++23, not C23, right? (You should update the patch title and summary accordingly.)

tbaederr commented on clang/test/CodeGen/constexpr-math.cpp:
```diff
@@ -0,0 +1,51 @@
+  constexpr double f14 = __builtin_frexpf(INFINITY, &i);
+
+  return 0;
+}
```
These all don't test the value of `i`, right?

zahiraam replied:
Right. The value of `i` is not added. Shoult it be added?

zahiraam commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -2922,7 +2922,7 @@ static bool handleFloatFloatBinOp(EvalInfo &Info, const BinaryOperator *E,
   //   mathematically defined [...], the behavior is undefined.
   // FIXME: C++ rules require us to not conform to IEEE 754 here.
-  if (LHS.isNaN()) {
+  if (!Info.getLangOpts().CPlusPlus23 && LHS.isNaN()) {
```
Not sure. We land here when the constexpr builtin_* is using a `NAN` argument with a Windows definition which is:
`#define INFINITY ((float)(1e+300 * 1e+300))`
`#define NAN      (-(float)(INFINITY * 0.0F))`

and an` EvalMode= EM_ConstantExpression`. 

`float f1 = __builtin_fminf(NAN, NAN);` returns an `EvaluateFloat() = true` (`EvalMode= EM_IgnoreSideEffects`)
`constexpr float f1 = __builtin_fminf(NAN, NAN);` returns an `EvaluateFloat() = false` (` EvalMode= EM_ConstantExpression`).

May be this is not the right place to do that? Should `keepEvaluatingAfterUndefinedBehavior` return true for C++23 when `EvalMode= EM_ConstantExpression`?

zahiraam replied:
May be additional testing should be added with Linux definition of `NAN`.

zahiraam replied:
This is basically to allow calls of __bultin_func(NAN, NAN) to be legal.

AaronBallman replied:
I think we're potentially misinterpreting the standard here regarding NANs. CC @jcranmer-intel @hubert-reinterpretcast 

https://eel.is/c++draft/basic.fundamental#13 is the interesting bit (I think), and I believe it's saying that if the floating-point representation you use has a way to represent a mathematical value then that value is "representable in that type". e.g., if your fp format can support positive/negative infinity, then a positive/negative infinity is considered representable. Note 10 says that more explicitly as: "Since negative and positive infinity are representable in ISO/IEC/IEEE 60559 formats, all real numbers lie within the range of representable values of a floating-point type adhering to ISO/IEC/IEEE 60559."

NANs are representable within ISO/IEC 60559 and are a mathematical result of an operation such as zero divided by zero. So I believe NANs are valid in a constant expression.

MSVC, Clang, and GCC all agree that they're not, but EDG believes they are: https://godbolt.org/z/cxPcKGvP3

AaronBallman replied:
The plot thickens, this may be a case where it's defined in C but C++ doesn't bother defining much of anything about floating-point and thus it is UB: https://github.com/cplusplus/draft/issues/5407

That said, https://eel.is/c++draft/basic.fundamental#12.sentence-8 seems to make this implementation-defined and it would be reasonable for us to define C++ as following the C semantics in situations where C++ is silent on the behavior.

C is more clear on this point. C23 5.2.5.3.3p8: Floating types shall be able to represent signed zeros or an unsigned zero and all normalized floating point numbers. In addition, floating types may be able to contain other kinds of floating-point numbers, such as subnormal floating-point numbers and unnormalized floating-point numbers, and
values that are not floating-point numbers, such as NaNs and (signed and unsigned) infinities.

cor3ntin replied:
This is a C function the semantics should reasonably be that of C
https://eel.is/c++draft/c.math#cmath.syn-1

fmin is well defined for Nan https://en.cppreference.com/w/c/numeric/math/fmin
frexp is well defined for NaN https://en.cppreference.com/w/cpp/numeric/math/frexp
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf

zahiraam replied:
So, the comment in the code above is not quite correct right? Should that entire condition be removed?

jcranmer-intel replied:
https://eel.is/c++draft/library.c#3 says that

>A call to a C standard library function is a non-constant library call ([defns.nonconst.libcall]) if it raises a floating-point exception other than FE_INEXACT.)"

which suggests that floating-point exceptions should generally cause things to fall out of constant expressions. C's Annex F suggests that warnings be emitted for constant expressions that cause floating-point exceptions other than FE_INEXACT (see F.8.2p2).

As for wording, [expr.pre]p4 requires that things be "mathematically defined", except IEEE 754 specifically says (in section 3.2)

> The mathematical structure underpinning the arithmetic in this standard is the extended reals, that is, the set
of real numbers together with positive and negative infinity.

The C++ specification actually does define "mathematically defined" somewhere, in a footnote of https://eel.is/c++draft/sf.cmath.general:

> A mathematical function is mathematically defined for a given set of argument values (a) if it is explicitly defined for that set of argument values, or (b) if its limiting value exists and does not depend on the direction of approach.

(which, as pendants will note, only defines it for functions, not the base operations of C++, and sidesteps the question of what the set of argument values actually is).

It does seem pretty clear from all the standards that a result of infinity is clearly part of the representable types, and it seems a reasonable inference that the result being infinity is "mathematically defined" (division by 0 is permissible in the extended real numbers).

The role of NaNs is... clear as mud. You have to go into the definition of <limits> to find out that NaN is a representable value. IEEE 754 gives multiple specification levels for floating-point, where NaN is a part of some of them, and proceeds to ignore the fact that it did so in the rest of the specification; the definition of operations involving NaN is deferred into an entire separate section. Section 7.2 goes so far as to say that

> The invalid operation exception is signaled if and only if there is no usefully definable result. In these cases
the operands are invalid for the operation to be performed.

Is `0.0 / 0.0` "mathematically defined" per IEEE 754? It's definitely *defined*, but I don't think I can entirely endorse saying that it's "mathematically defined." The rule of thumb that FE_INVALID is signaled when the inputs are non-qNaN but the output is NaN, combined with the above text, makes it suspect that it should be considered a "mathematical" definition. But this also doesn't help with `NaN / NaN`, which is `NaN` but doesn't throw a FE_INVALID.

As noted in the CWG issue Aaron linked to, someone needs to write a paper to clarify this, and... that's going to be me isn't it?

hubert-reinterpretcast replied:
My understanding of the intended design was that operations producing NaNs from non-NaNs in C++ constexpr cause the expression to not be constant. That said, NaNs are still in play because of `numeric_limits`.

For `frexp` though, I think there being an "unspecified" value stored for the exponent is a problem where the design could use more clarity. I would think that the operation could still be performed, but the "unspecified" value should probably be considered indeterminate.

cor3ntin replied:
> My understanding of the intended design was that operations producing NaNs from non-NaNs in C++

What is "operation" here? NaN in arithmetics / conversions / other core language etc are one thing, but
when calling a function which is specifically specified to produce NaN in some cases... it's very unclear to me that the program should be invalid.

If that's the intent, is there anywhere where it would be documented?

frederick-vs-ja replied:
There's an open CWG issue ([CWG2168](https://cplusplus.github.io/CWG/issues/2168.html)) about this.

zahiraam replied:
> There's an open CWG issue ([CWG2168](https://cplusplus.github.io/CWG/issues/2168.html)) about this.

That's referring to INFINITY though. This is more specifically about NAN.

hubert-reinterpretcast replied:
> Do we want to do the same for C++23? If not what's exactly an "unspecified" value?

It means that `INT_MAX` and `INT_MIN` is fine (based on the wording). The committee discussion had already decided that the constexpr evaluation need not match the target's runtime in terms of implementation-defined behavior with respect to:
- what cases are errors, and
- how non-error cases of NaNs and infinities, etc. are handled.

Instead, C's Annex F was to specify the behaviour; however, that C's Annex F did not fully specify the result (as in the case in question) was not discussed.

hubert-reinterpretcast replied:
I do not think there is any particular reason why this change is C++23 specific.
The question remains whether the change should be made at all.

If the primary rationale here is to facilitate testing using the Windows header definitions, I suspect the answer is that the Windows definition of NaN is not a C++ constant expression.

@jcranmer-intel's comment (https://github.com/llvm/llvm-project/pull/88978/files#discussion_r1576817678), supports the interpretation that operations that result in NaN from non-qNaN inputs are not valid in a constant expression.

Other than wanting to use the Windows header definitions for the tests here, is there a reason why this change belongs in this PR?

zahiraam replied:
Removed and edit the SemaCXX/constexpr-math.cpp to reflect the change.

zahiraam committed:
Addressed review comments.

jcranmer-intel reviewed (commented):
Other side notes:

fmin and frexp can signal exceptions if the input is an sNaN, which causes [library.c]p3 to kick in. (That's the only time these operations can signal an exception.)

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,57 @@
+    assert(__builtin_iszero(__builtin_fminf(+INFINITY, 0)));
+
+    // frexp
+    static_assert(__builtin_frexp(123.45, &i) == 0.96445312500000002);
```
This should not be a constant expression. This line is not supposed to compile.
`i` is supposed to be written to, but it lives outside of the constant expression evaluation.

zahiraam replied:
I am reading from `https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0533r9.pdf` that the return type of __builtin_frexp() can be constexpr. 
This is the result of the code I have added in VisitCallExpr (`case Builtin::BI__builtin_frexpf:
case Builtin::BI__builtin_frexp: {}  that probably needs to be guarded by getLangOpts().CPlusPlus23).

hubert-reinterpretcast replied:
Just because a function is `constexpr` does not mean an invocation of it would be allowed in a constant expression even if the arguments were constant expressions.

The specific issue is that the semantics of `frexp` has it write a value through it's second parameter. In order to modify the stored value of an object in constant evaluation, the object has to begin its lifetime in that constant evaluation.

frederick-vs-ja replied:
It seems that we should write something like
```C++
struct { double num_; int expo_; } result;
result.num_ = frexp(x, &result.expo_);
return result;
```

zahiraam replied:
Still working on this. Can you please review the implementation of frexp I have added in `ExprConstant.cpp`?
@frederick-vs-ja  are you suggesting using something like this for the use of `frexp` in this test?
```
struct FrexpResult {
  double fraction;
  int exponent;
};

FrexpResult simulateFrexp(double value) {
  FrexpResult result;
  result.fraction = __builtin_frexp(value, &result.exponent);
  return result;
}

int main () {
  double value = 123.456;
  FrexpResult result = simulateFrexp(123.456);
  
  return 0;
}

```
Still no way of asserting what the return values of fraction and exponent are from calls to `simulateFrex`p with various arguments. Any suggestions?

hubert-reinterpretcast replied:
@zahiraam, make `simulateFrexp` a `constexpr` function (maybe with `float` and `long double` overloads as well).

```
template <typename T>
struct FrexpResult {
  constexpr FrexpResult(T fraction, int exponent) : fraction(fraction), exponent(exponent) {}
  T fraction;
  int exponent;
  friend constexpr bool operator<=>(const FrexpResult &, const FrexpResult &) = default;
};

constexpr auto q() { return FrexpResult{0.5, 0}; };
static_assert(q() == FrexpResult{0.5, 0});
```

hubert-reinterpretcast replied:
Here is an example of checking both the fraction and exponent values by "identity" instead of by comparisons.

```cpp
template <auto, auto> constexpr bool is_same_val = false;
template <auto X> constexpr bool is_same_val<X, X> = true;

template <typename T>
struct FrexpResult {
  T fraction;
  int exponent;
};

template <typename T>
constexpr auto simulateFrexp(T value) {
  FrexpResult<T> result;
  result.fraction = __builtin_frexp(value, &result.exponent);
  return result;
}

static_assert(is_same_val<simulateFrexp(1.), FrexpResult{0.5, 1}>);
```

https://godbolt.org/z/Whx6o5Y7E

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14547,6 +14547,17 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
+        !EvaluatePointer(E->getArg(1), Pointer, Info))
+      return false;
+    llvm::RoundingMode RM = getActiveRoundingMode(Info, E);
+    int FrexpExp;
+    Result = llvm::frexp(Result, FrexpExp, RM);
```
The value in `FrexpExp` is just ignored?

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14547,6 +14547,17 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
     return false;
 
+  case Builtin::BI__builtin_frexpf:
+  case Builtin::BI__builtin_frexp: {
```
Why no `long double`?

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,55 @@
+{
+  int i;
+
+  // fmin
```
No testing for `long double`?

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14638,6 +14649,8 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
     return true;
   }
 
+  case Builtin::BIfmin:
+  case Builtin::BIfminf:
```
Don't make these evaluate in `constexpr`. Not only does this extend C (not a stated intent of the patch), it will cause accidental dependencies that break when `-fno-builtin` is used.

philnik777 replied:
These have to be made `constexpr`, at least in C++23. C++ stdlibs can't override them to make them call the `__builtin_` versions, so Clang has to handle that.

hubert-reinterpretcast replied:
Makes me wonder how far C++ stdlibs will get if they included the C library headers in an inline namespace under the global namespace (and count on `extern "C"` to handle the linkage names).

philnik777 replied:
How would that help?

hubert-reinterpretcast replied:
> How would that help?

I guess it doesn't... The functions would be ambiguous when called...

hubert-reinterpretcast replied:
I think we'll be missing at least "pedantic" diagnostics if we don't restrict the `constexpr`-ness to C++23 and up. So we do need that restriction.

The interaction with `-fno-builtin[-*]` may need more discussion. At least the documentation for that option would need to be updated if the C++23 `constexpr` math would break with it.

hubert-reinterpretcast replied:
At the Clang C/C++ Language Workgroup call on 2024-05-15, it was agreed that `-fno-builtin[-*]` should remain capable of disabling built-in treatment even when that treatment would be required for conformance w.r.t. `constexpr` behaviour of functions shared with C.

For me, the first expectation for this PR (on that subject) is an update to the documentation of `-fno-builtin[-*]`.

@AaronBallman @ldionne, I am no sure how we want to handle this w.r.t. documenting the C++23 implementation status. Does it go in both the Clang documentation (for the functions shared with C) and the libc++ documentation (for the overloads added by C++)?

philnik777 replied:
I don't think anything should happen to the libc++ documentation in this patch. We can update the implementation to behave the same across the different overloads in a follow-up and update our documentation then.

hubert-reinterpretcast commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -3440,7 +3440,7 @@ def Fmod : FPMathTemplate, LibBuiltin<"math.h"> {
-  let Attributes = [NoThrow];
+  let Attributes = [NoThrow, Constexpr];
   let Prototype = "T(T, int*)";
   let AddBuiltinPrefixedAlias = 1;
```
Should probably match `Fmin` on this:
```suggestion
  let AddBuiltinPrefixedAlias = 1;
  let OnlyBuiltinPrefixedAliasIsConstexpr = 1;
```

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,57 @@
+    static_assert(__builtin_fmin(+0.0, -0.0) == -0.0, "");
+    assert(__builtin_isnan(__builtin_fminf(NAN,NAN)));
+    assert(__builtin_isnan(__builtin_fminf(NAN, -1)));
+    assert(__builtin_isnan(__builtin_fminf(-INFINITY, 0)));
```
I doubt the runtime assertions are working (or appropriate in the Clang LIT testing).

zahiraam committed:
Responded to review comments.

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14547,6 +14547,20 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
+      return false;
+    llvm::RoundingMode RM = getActiveRoundingMode(Info, E);
+    int FrexpExp;
+    FrexpExp = ilogb(Result);
+    FrexpExp = FrexpExp == llvm::detail::IEEEFloat::IEK_Zero ? 0 : FrexpExp + 1;
+    Result =  scalbn(Result, -FrexpExp, RM);
```
This is not an improvement over
https://github.com/llvm/llvm-project/blob/d7050b53a75a23024502a9e4f56b385256c4722f/clang/lib/AST/ExprConstant.cpp#L14558

It is still the case that no attempt is made to store the value of `FrexpExp` via `Pointer`.

Additionally, (although potentially "better" than the status quo of `llvm::frexp`) this adds yet another place where the calculation for IBM double-double would potentially need special maintenance. See https://github.com/llvm/llvm-project/issues/90287.

zahiraam replied:
Added function` StoreExponent`.

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,59 @@
+// CHECK: store double 1.300000e+00, ptr {{.*}}
+// CHECK: store double -0.000000e+00, ptr {{.*}}
+// CHECK: store double -0.000000e+00, ptr {{.*}}
+// WIN: store float 0xFFF8000000000000, ptr {{.*}}
+// LNX: store float 0x7FF8000000000000, ptr {{.*}}
+// CHECK: store float -1.000000e+00, ptr {{.*}}
+// CHECK: store float 0xFFF0000000000000, ptr {{.*}}
```
This only checks three of the four `float` values. Maybe the test can use `CHECK-NEXT`, `WIN-NEXT`, etc.?

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,59 @@
+// RUN: %clang_cc1 -x c++ -triple x86_64-unknown-unknown -std=c++23 \
+// RUN: -DWIN -emit-llvm -o - %s | FileCheck %s --check-prefixes=WIN
+
+// RUN: %clang_cc1 -x c++ -triple x86_64-unknown-unknown -std=c++23 \
+// RUN: -emit-llvm -o - %s | FileCheck %s --check-prefixes=LNX
```
None of the `CHECK` lines were being checked...
```suggestion
// RUN: -DWIN -emit-llvm -o - %s | FileCheck %s --check-prefixes=CHECK,WIN

// RUN: %clang_cc1 -x c++ -triple x86_64-unknown-unknown -std=c++23 \
// RUN: -emit-llvm -o - %s | FileCheck %s --check-prefixes=CHECK,LNX
```

zahiraam replied:
If you don't think it's too verbose, I have used the `update_cc_test_checks.p`y tool to generate the IR.

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,59 @@
+  constexpr double f9 = __builtin_frexp(123.45, &i);
+  constexpr double f10 = __builtin_frexp(0.0, &i);
+  constexpr double f11 = __builtin_frexp(-0.0, &i);
+  constexpr double f12 = __builtin_frexpf(NAN, &i);
+  constexpr double f13 = __builtin_frexpf(-NAN, &i);
+  constexpr double f14 = __builtin_frexpf(INFINITY, &i);
```
Copy/paste errors?
```suggestion
  constexpr float f12 = __builtin_frexpf(NAN, &i);
  constexpr float f13 = __builtin_frexpf(-NAN, &i);
  constexpr float f14 = __builtin_frexpf(INFINITY, &i);
```

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,38 @@
+
+    // frexp
+    static_assert(__builtin_frexp(123.45, &i) == 0.96445312500000002);
+    static_assert(!__builtin_isnan(__builtin_frexp(123.45, &i)), "");
```
Why is this necessary?
```suggestion
```

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,38 @@
+#define NAN (__builtin_nanf(""))
+#define INFINITY (__builtin_inff())
+#endif
+
+int main() {
```
Use identity instead of comparisons/classifications for validation:
```suggestion

template <auto, auto> constexpr bool is_same_val = false;
template <auto X> constexpr bool is_same_val<X, X> = true;

int main() {
```

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,38 @@
+
+    // fmin
+    static_assert(__builtin_fmin(15.24, 1.3) == 1.3, "");
+    static_assert(__builtin_fmin(-0.0, +0.0) == -0.0, "");
+    static_assert(__builtin_fmin(+0.0, -0.0) == -0.0, "");
```
... also, optional `static_assert` message should be removed if not used in a helpful manner.

Example (please apply to other lines too):
```suggestion
    static_assert(is_same_val<__builtin_fmin(-0.0, +0.0), -0.0>);
    static_assert(is_same_val<__builtin_fmin(+0.0, -0.0), -0.0>);
```

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,38 @@
+    static_assert(__builtin_isnan(__builtin_fminf(NAN,NAN)), "");
+
+    // frexp
+    static_assert(__builtin_frexp(123.45, &i) == 0.96445312500000002);
```
Use `123.45 / 128.` instead of `0.96445312500000002`.

zahiraam committed:
Added code to store the exponent in Pointer. Edited LIT tests
as requested by reviewer.

zahiraam commented:
@hubert-reinterpretcast Thanks for the reviews/suggestions. Let me know what you think.

zahiraam committed:
Added fmax function and function IsinRange function.

hubert-reinterpretcast commented on llvm/include/llvm/ADT/APFloat.h:
```diff
@@ -951,6 +951,19 @@ class APFloat : public APFloatBase {
   bool needsCleanup() const { APFLOAT_DISPATCH_ON_SEMANTICS(needsCleanup()); }
 
+  //  Checks that the value x is in the range (-1;-0.5], [0.5; 1)
+  static bool isInRange(const llvm::APFloat &x) {
```
I don't think this should be a member function of `APFloat`. It can be an implementation detail of `ExprConstant.cpp`. In either case, the function should be named something like `isInFrexpResultRange`.

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14572,6 +14572,9 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
     int FrexpExp;
     llvm::RoundingMode RM = getActiveRoundingMode(Info, E);
     Result = llvm::frexp(Result, FrexpExp, RM);
+    if (!Result.isZero() && !Result.isNaN() && !Result.isInfinity())
+      assert(llvm::APFloat::isInRange(Result) &&
+            "The value is not in the expected range for frexp.");
```
assert((Result.isZero() || Result.isNaN() || Result.isInfinity() || ...) && "...");

hubert-reinterpretcast reviewed (commented):
Partial review comments.

zahiraam committed:
Addressed review comments.

zahiraam committed:
Added a fix for failing LIT test.

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14574,9 +14574,17 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
   case Builtin::BI__builtin_frexp:
-  case Builtin::BI__builtin_frexpf:
-  case Builtin::BI__builtin_frexpl: {
+  case Builtin::BI__builtin_frexpf: {
```
The comment is incorrect. The _actual_ AIX library function `frexpl` has (some version of) `PPCDoubleDouble` type. Because of that, Clang (which only supports 64-bit `long double` on AIX) has to map calls to `__builtin_frexpl` to `frexp`.

For compile-time evaluation, the `APFloat`s associated with `long double` on AIX have the IEEE double format (and the evaluation works fine).

The solution to the failing test is to change it. The problem with the test is that the first input argument is a constant; changing it to refer to a global variable should work.
https://github.com/llvm/llvm-project/blob/566fbb450092bf8c9f966a6ab1b0381626e3e535/clang/test/CodeGen/aix-builtin-mapping.c#L16

AaronBallman replied:
> For compile-time evaluation, the APFloats associated with long double on AIX have the IEEE double format (and the evaluation works fine).

Does the evaluation produce the same results as would happen at runtime despite the different format?

hubert-reinterpretcast replied:
> Does the evaluation produce the same results as would happen at runtime despite the different format?

There is no different format. The default format for the `long double` type on AIX (and the only one supported by Clang for AIX) is IEEE double.

The IR generated for basic `long double` operations might clarify: https://godbolt.org/z/W4jqr6Kva

hubert-reinterpretcast replied:
> With this change (special casing frexpl for AIX) the IR produced is: `call { double, i32 } @llvm.frexp.f64.i32(double 1.234500e+02)`
> 
> without this change the IR is: `store double 0.000000e+00, ptr %returnValue, align 8`
> 
> This doesn't seem correct to me.

Ignoring considerations around storing to `DummyInt`, the original test code was:
```
  returnValue = __builtin_frexpl(0.0L, &DummyInt);
```
I don't know where the `123.45` came from. That the result is zero for an input of zero is correct.

hubert-reinterpretcast replied:
> I was referring to https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGBuiltin.cpp#L3543.

That affects the IR generated by Clang when Clang _does not_ constant-evaluate (and was motivated by backend issues).

zahiraam replied:
> > With this change (special casing frexpl for AIX) the IR produced is: `call { double, i32 } @llvm.frexp.f64.i32(double 1.234500e+02)`
> > without this change the IR is: `store double 0.000000e+00, ptr %returnValue, align 8`
> > This doesn't seem correct to me.
> 
> Ignoring considerations around storing to `DummyInt`, the original test code was:
> 
> ```
>   returnValue = __builtin_frexpl(0.0L, &DummyInt);
> ```
> 
> I don't know where the `123.45` came from. That the result is zero for an input of zero is correct; however, the constant-folding in the context of that test would undermine its intent (see check that an appropriate target runtime function is called).

Sorry. The output is:   `call { double, i32 } @llvm.frexp.f64.i32(double 0.000000e+00)`. That was a mistake of my part.

zahiraam committed:
Fixed aix-builtin-mapping.c test.

zahiraam committed:
Merge remote-tracking branch 'origin/main' into C23ConstExpr

hubert-reinterpretcast commented:
Re: https://github.com/llvm/llvm-project/pull/88978#discussion_r1578762448

> It means that INT_MAX and INT_MIN is fine (based on the wording).

The committee reflector discussion reached a conclusion that this behaviour is okay. Implementations need not agree on the values. Indeed, an implementation is not required to consistently produce the same value.

hubert-reinterpretcast commented:
> Some more thought is needed on how to handle the non-`__builtin_`-prefixed cases under non-C++23-or-higher language modes. The specific implications of those functions being non-`constexpr` under said modes (as required, for C++, by https://wg21.link/constexpr.functions) may determine the solution to apply in this PR.

It seems that Clang _does_ care whether or not a function is `constexpr` (outside of an actual attempt to evaluate it during constant evaluation) when it is trying to diagnose the pre-C++23 IFNDR situation of a `constexpr` function that never produces a constant expression: https://godbolt.org/z/hnfqrrrdM

```cpp
extern "C" double fmin(double, double); // expected-note {{declared here}}
constexpr double contrived() { return fmin(0., 0.); }
// expected-error@-1 {{constexpr function never produces a constant expression}}
// expected-note@-2 {{non-constexpr function 'fmin' cannot be used in a constant expression}}
```

Aside: Clang checks the arguments before the function declaration.

hubert-reinterpretcast commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -3452,9 +3452,10 @@ def Fmod : FPMathTemplate, LibBuiltin<"math.h"> {
 def Frexp : FPMathTemplate, LibBuiltin<"math.h"> {
   let Spellings = ["frexp"];
-  let Attributes = [NoThrow];
+  let Attributes = [NoThrow, Constexpr];
```
The `Constexpr` here makes the non-`__builtin_`-prefixed version of the function `constexpr` under all language levels.
I checked the code: The `OnlyBuiltinPrefixedAliasIsConstexpr` flag can _add_ `constexpr`-ness; it does not _restrict_ it.

For reference:
https://github.com/llvm/llvm-project/blob/702198fc9ac5dba392f9d9ba7c56467996343c0a/clang/utils/TableGen/ClangBuiltinsEmitter.cpp#L225-L226

We may need a new builtin attribute that expresses `constexpr` for the non-prefixed functions only for C++23 and up (but we might end up with C++26-and-up, etc. in the future). However, it seems making the non-prefixed functions actually non-`constexpr` (as opposed to simply failing to evaluate) is only a QoI concern (see https://github.com/llvm/llvm-project/pull/88978#issuecomment-2119080104 for more context).

@philnik777, thoughts?

philnik777 replied:
This also influences `__has_constexpr_builtin(fmin)`, which would return true but then fail to actually evaluate during a constant expression, which doesn't seem that great. IMO we'd probably want to introduce the concept of "constexpr since C++xy".

hubert-reinterpretcast replied:
> IMO we'd probably want to introduce the concept of "constexpr since C++xy".

@philnik777, are you envisioning something where `LangOptions` are passed into `Builtin::Context::isConstantEvaluated`?
https://github.com/llvm/llvm-project/blob/fb2c6597e39e9e1a775525ea0236b2f89e46acff/clang/include/clang/Basic/Builtins.h#L279

I think the existing `Constexpr`/"`E`" attribute can be left alone for use in "always `constexpr`" cases.

For the selectively-`constexpr` cases, `LibBuiltin` in `clang/include/clang/Basic/BuiltinsBase.td` should have `OnlyBuiltinPrefixedAliasIsConstexpr` renamed to `BuiltinPrefixedAliasIsConstexpr`. Then adding a new string property for `UnprefixedOnlyConstexprSince` makes sense to me. It can be encoded into `Builtin::Info` (and I think using the `LangFeatures` enumeration from `"clang/Basic/LangStandard.h"` is not terribly wrong).

zahiraam replied:
Done.

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14683,6 +14710,23 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
 
+  case Builtin::BI__builtin_frexp:
+  case Builtin::BI__builtin_frexpf:
+  case Builtin::BI__builtin_frexpl: {
```
`__builtin_frexpf16` and `__builtin_frexp128` should probably be added at the same time as the other `__builtin_*` cases.

zahiraam replied:
You mean adding :
`case Builtin::BI__builtin_frexp16:`
`case Builtin::BI_builtin_frexp128:`
to the cases here?

zahiraam replied:
Done.

hubert-reinterpretcast commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -3452,9 +3452,10 @@ def Fmod : FPMathTemplate, LibBuiltin<"math.h"> {
 
 def Frexp : FPMathTemplate, LibBuiltin<"math.h"> {
```
There is a separate `FrexpF16F128` that should probably also be updated.

zahiraam replied:
I have added `UnprefixedOnlyConstexprSince` to the class `LIBBUILTIN` so it will be unknow if I add it to this def. I will wait to see if the changes I made were correct and update this rule accordingly.

hubert-reinterpretcast commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14674,6 +14676,31 @@ static bool TryEvaluateBuiltinNaN(const ASTContext &Context,
+    clang::IntegerLiteral *IL =
+        clang::IntegerLiteral::Create(Info.Ctx, llvm::APSInt(32, exp),
+                                      Info.Ctx.IntTy, clang::SourceLocation());
+    VarD->setInit(IL);
```
See use of `handleAssignment` for `__builtin_add_overflow`:
https://github.com/llvm/llvm-project/blob/b2d7d72ff2408dcc27048e08d50ac719ff03a89e/clang/lib/AST/ExprConstant.cpp#L13071

zahiraam replied:
Not really sure what I should be looking at here. Are you objecting to the way StoreExponent is written?

hubert-reinterpretcast replied:
> Not really sure what I should be looking at here. Are you objecting to the way StoreExponent is written?

Sorry for the late response. Yes, I am objecting to the way `StoreExponent()` is written. It tries to change the initializer of a variable to give it a value, but that approach is wrong. There does not need to be a variable in the first place (the pointer can point to a member of a temporary class object). Please see `handleAssignment()`.

zahiraam replied:
Done.

hubert-reinterpretcast commented on clang/test/CodeGen/aix-builtin-mapping.c:
```diff
@@ -6,17 +6,21 @@
 
   returnValue = __builtin_modfl(1.0L, &DummyLongDouble);
-  returnValue = __builtin_frexpl(0.0L, &DummyInt);
+  returnValue = __builtin_frexpl(input, &DummyInt);
```
Once the handling of the second parameter is corrected, the change to the test should become unnecessary.

zahiraam committed:
Addressed review comments.

zahiraam commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,52 @@
+// RUN: %clang_cc1 -DWIN -verify -std=c++23 -fsyntax-only  %s
+// RUN: %clang_cc1 -verify -std=c++23 -fsyntax-only  %s
```
Done.

zahiraam commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14683,6 +14710,23 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
   default:
     return false;
 
+  case Builtin::BI__builtin_frexp:
```
Added the non-`__builtin_ prefixed cases.

zahiraam committed:
Put back CodeGen/aix-builtin-mapping.c how it was originally written.

hubert-reinterpretcast commented:
> All the LIT tests failing are due to the latest changes I made. I will edit them once I know that the latest implementation is correct.

Thanks. I may be delayed in reviewing this as the C++ committee meeting is the week after next.

zahiraam commented:
> > All the LIT tests failing are due to the latest changes I made. I will edit them once I know that the latest implementation is correct.
> 
> Thanks. I may be delayed in reviewing this as the C++ committee meeting is the week after next.

Ping @hubert-reinterpretcast

zahiraam commented:
@hubert-reinterpretcast  Would really appreciate a review of this please. Thanks.

jcranmer-intel commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14674,6 +14676,31 @@ static bool TryEvaluateBuiltinNaN(const ASTContext &Context,
 }
+//  Checks that the value x is in the range (-1;-0.5], [0.5; 1)
+static bool isInFrexpResultRange(const llvm::APFloat &x) {
+  llvm::APFloat minusOne(x.getSemantics(), "-1.0");
+  llvm::APFloat minusHalf(x.getSemantics(), "-0.5");
+  llvm::APFloat half(x.getSemantics(), "0.5");
+  llvm::APFloat one(x.getSemantics(), "1.0");
+
+  return ((x.compare(minusOne) == llvm::APFloat::cmpGreaterThan &&
+           x.compare(minusHalf) != llvm::APFloat::cmpGreaterThan) ||
+          (x.compare(half) != llvm::APFloat::cmpLessThan &&
+           x.compare(one) == llvm::APFloat::cmpLessThan));
```
You can simplify this by checking that the absolute value is in [0.5, 1)

zahiraam committed:
Fix build issue in flang.

zahiraam committed:
Merge remote-tracking branch 'origin/main' into C23ConstExpr

zahiraam committed:
Merge remote-tracking branch 'origin/main' into C23ConstExpr

zahiraam committed:
Rewrote isInFrexpResultRange according to review comments and fixed
format.

zahiraam committed:
Fix format.

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,162 @@
+  constexpr double frexp3 = __builtin_frexp(-0.0, &i);
+  constexpr float frexp4 = __builtin_frexpf(NAN, &i);
+  constexpr float frexp5 = __builtin_frexpf(-NAN, &i);
+  constexpr float frexp6 = __builtin_frexpf(INFINITY, &i);
+  constexpr float frexp7 = __builtin_frexpf(INFINITY, &i);
```
@zahiraam, this comment seems to be not addressed?

zahiraam committed:
Addressed missed review comments.

hubert-reinterpretcast commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -253,8 +253,10 @@ def FmodF16F128 : F16F128MathTemplate, Builtin {
 def FrexpF16F128 : F16F128MathTemplate, Builtin {
   let Spellings = ["__builtin_frexp"];
-  let Attributes = [FunctionWithBuiltinPrefix, NoThrow];
+  let Attributes = [FunctionWithBuiltinPrefix, NoThrow, Const, ConstexprSince];
```
The `frexp` family of functions cannot be marked `const` because they modify the object pointed to by the pointer parameter.

zahiraam replied:
Done.

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,106 @@
+  constexpr float frexp6 = __builtin_frexpf(+INFINITY, &i);
+  constexpr float frexp7 = __builtin_frexpf(-INFINITY, &i);
+  constexpr long double frexp8 = __builtin_frexpl(259.328L, &i);
+  constexpr float frexp9 = __builtin_frexpf16(3.5, &i);
+  constexpr long double frexp10 = __builtin_frexpf128(234.78, &i);
```
```suggestion
  constexpr __fp16 frexp9 = __builtin_frexpf16(3.5, &i);
  constexpr __float128 frexp10 = __builtin_frexpf128(234.78, &i);
```

hubert-reinterpretcast commented on clang/test/CodeGenCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,106 @@
+  constexpr long double max8 = __builtin_fmaxl(123.456L, 789.012L);
+
+  // frexp
+  constexpr double frexp1 = __builtin_frexp(123.45, &i);
```
@zahiraam, to be explicit, these tests (as-is) will fail once the comment re: `StoreExponent()` is addressed.

This will probably work:
```suggestion
  constexpr double frexp1 = __builtin_frexp(123.45, (int [1]){});
```

zahiraam replied:
Done.

hubert-reinterpretcast commented on clang/test/SemaCXX/constexpr-math.cpp:
```diff
@@ -0,0 +1,108 @@
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+  static_assert(is_same_val<CALL_BUILTIN(frexpl, 123.45L, &i), 123.45L/128>);
+  // cplusplus20andless-error@-1 {{non-type template argument is not a constant expression}}
+
```
Add the `__fp16` and `__float128` cases too.

zahiraam replied:
Done.

zahiraam committed:
Addressed review comments. Remove StoreExponent function and fixed LIT
tests.

zahiraam commented:
Ping?

hubert-reinterpretcast commented:
> Ping?

The changes to add the "constexpr since" property are missing changes to the tablegen utility. Figuring out how it would work is more involved than I had the time for right now.

zahiraam commented:
> Also, if at all possible, I think we only want the extra field for "library builtins" and not all builtins.

Yes, I have tried limiting it to `LibBuiltin` but the addition of `FrexpF16F128` led me to add it `Builtin` too.

> Alternatively, we don't need tablegen changes if we leave the encoding of which language standard a builtin is constexpr since to a helper function.

Will look into this.

Thanks.

hubert-reinterpretcast commented:
> Yes, I have tried limiting it to `LibBuiltin` but the addition of `FrexpF16F128` led me to add it `Builtin` too.

I believe that FrexpF16F128 only has the `__builtin_`-prefixed versions. That would mean that it does not need to be "constexpr since" (just plain "constexpr").

zahiraam committed:
Replaced ConstexpSince by Constexpr for FrexpF16F128.

zahiraam committed:
Removed builtin attribute ConstExprSince and replaced it with a member
function.

zahiraam commented:
> Alternatively, we don't need tablegen changes if we leave the encoding of which language standard a builtin is constexpr since to a helper function.

@hubert-reinterpretcast I have implemented what you proposed. Removed the `ConstExprSince` attribute and replaced it with member function. Is that what you were thinking of?

zahiraam committed:
Fix format.

zahiraam committed:
Fix LIT failures.

zahiraam committed:
Merge remote-tracking branch 'origin/main' into C23ConstExpr

zahiraam commented:
ping @hubert-reinterpretcast

cor3ntin commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14951,6 +14961,36 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
+  case Builtin::BI__builtin_frexpl:
+  case Builtin::BI__builtin_frexpf16:
+  case Builtin::BI__builtin_frexpf128: {
+    const auto *FDecl = E->getDirectCallee();
+    clang::LangStandard::Kind ConstExprSinceVersion =
+        FDecl->getConstexprBuiltinSinceVersion();
+    if (ConstExprSinceVersion > Info.Ctx.getLangOpts().LangStd)
+      return false;
+    LValue Pointer;
+    if (!EvaluateFloat(E->getArg(0), Result, Info) ||
+        !EvaluatePointer(E->getArg(1), Pointer, Info))
```
I think the builtin should be unconditionally constexpr in Clang. 

And we should certainly not store additional info for  that in `FunctionDecl`, which has a huge cost in terms of AST size.

Instead, the builtin should become constexpr, and libc++  can decide not to make the wrapper constexpr in older language modes, or to have some availability warning in older language modes.

cor3ntin replied:
@hubert-reinterpretcast I think if we wanted to do something about the non-prefixed builtin, I would prefer
 - that we table-gen the information
 - that the version is encoded as an attribute (it might just be an implementation detail attribute) - so as to not bloat the AST
 - that we make it a diagnostic that users can turn into an error or not (and then we can discuss what the default should be)
